pkg/apis/resource/helper.go

```diff
@@ -1,11 +1,187 @@
+	resourceReg = regexp.MustCompile(`\${res\.([^.}]+)\.([^.}]+)}`)
+
+	// VariableFormat the format of variable.
+	variableFormat = `"${var.%s}"`
+	// ResourceFormat the format of resource output.
+	resourceFormat = `"${res.%s.%s}"`
+)
+
+func injectAttributes(
+	ctx context.Context,
+	client model.ClientSet,
+	projectID, environmentID object.ID,
+	attrs property.Values,
+) (property.Values, error) {
+	bs, err := json.Marshal(attrs)
+	if err != nil {
+		return nil, err
+	}
+
+	varNames, resNames, resOuts := getInterpolationTargets(bs)
+
+	if len(varNames) == 0 && len(resNames) == 0 {
+		return attrs, nil
+	}
+
+	varValues, err := getInterpolationVariableValues(ctx, client, varNames, projectID, environmentID)
+	if err != nil {
+		return nil, err
+	}
+
+	outputValues, err := getInterpolationOutputValues(ctx, client, resNames, projectID, environmentID, resOuts)
+	if err != nil {
+		return nil, err
+	}
+
+	return replaceAttributes(bs, varValues, outputValues)
+}
+
+func replaceAttributes(
+	attrByte []byte,
+	varValues map[string]json.RawMessage,
+	outputValues map[string]property.Value,
+) (property.Values, error) {
+	for n, v := range varValues {
+		attrByte = bytes.ReplaceAll(attrByte, []byte(n), v)
+	}
+
+	for n, v := range outputValues {
+		attrByte = bytes.ReplaceAll(attrByte, []byte(n), v)
+	}
+
+	var injectAttrs property.Values
+	if err := json.Unmarshal(attrByte, &injectAttrs); err != nil {
+		return nil, err
+	}
+
+	return injectAttrs, nil
+}
+
+// getInterpolationVariableValues returns should inject variable values.
+// Output is map of ${var.{variable_name}}:{value}.
+func getInterpolationVariableValues(
+	ctx context.Context,
+	client model.ClientSet,
+	varNames []string,
+	projectID, environmentID object.ID,
+) (map[string]json.RawMessage, error) {
+	// Get should inject variables.
+	vars, err := terraform.GetVariables(ctx, client, varNames, projectID, environmentID)
+	if err != nil {
+		return nil, err
+	}
+
+	values := make(map[string]json.RawMessage)
+
+	for _, v := range vars {
+		n := fmt.Sprintf(variableFormat, v.Name)
+		values[n] = []byte(fmt.Sprintf(`"%s"`, string(v.Value)))
+	}
+
+	return values, nil
+}
+
+// getInterpolationOutputValues returns should inject output values.
+// Input resOutputs is map of {resource_name}_{output_name}:{inject_type}.
+// Output is map of ${var.{resource_name}.{output_name}}:{value}.
+func getInterpolationOutputValues(
+	ctx context.Context,
+	client model.ClientSet,
+	resNames []string,
+	projectID, environmentID object.ID,
+	resOutputs map[string]string,
+) (map[string]property.Value, error) {
+	depIDs, err := client.Resources().Query().
+		Where(
+			resource.NameIn(resNames...),
+			resource.ProjectID(projectID),
+			resource.EnvironmentID(environmentID)).
+		IDs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get should inject resource outputs.
+	outputs, err := terraform.GetServiceDependencyOutputs(ctx, client, depIDs, resOutputs)
```

Comment:
Calling deployer/terraform pkg in apis/resource seems to be odd. Should we move the function to pkg/resource(and change its name)?